runtime/common/slot_range_helper/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! A helper macro for generating SlotRange enum.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use sp_std::{result, ops::Add, convert::TryInto};
+pub use sp_runtime::traits::CheckedSub;
+pub use parity_scale_codec::{Encode, Decode};
+
+#[macro_export]
+macro_rules! generate_slot_range{
+	// Entry point
+	($( $x:ident ( $e:expr ) ),*) => {
+		$crate::generate_slot_range!(@
+			{ }
+			$( $x ( $e ) )*
+		);
+	};
+	// Does the magic...
+	(@
+		{ $( $parsed:ident ( $t1:expr, $t2:expr ) )* }
+		$current:ident ( $ce:expr )
+		$( $remaining:ident ( $re:expr ) )*
+	) => {
+		paste::paste! {
+			$crate::generate_slot_range!(@
+				{
+					$( $parsed ( $t1, $t2 ) )*
+					[< $current $current >] ( $ce, $ce )
+					$( [< $current $remaining >] ($ce, $re) )*
+				}
+				$( $remaining ( $re ) )*
+			);
+		}
+	};
+	(@
+		{ $( $parsed:ident ( $t1:expr, $t2:expr ) )* }
+	) => {
+		$crate::generate_slot_range_enum!(@ $( $parsed )* );
+
+		$crate::generate_slot_range_count!(@ $( $parsed )* );
+
+		#[cfg(feature = "std")]
+		impl std::fmt::Debug for SlotRange {
+			fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+				let p = self.as_pair();
+				write!(fmt, "[{}..{}]", p.0, p.1)
+			}
+		}
+
+		impl SlotRange {
+			$crate::generate_slot_range_as_pair!(@ $( $parsed ( $t1, $t2 ) )* );
+
+			$crate::generate_slot_range_len!(@ $( $parsed ( $t1, $t2 ) )* );
+
+			$crate::generate_slot_range_new_bounded!(@ $( $parsed ( $t1, $t2 ) )* );
+		}
+	};
+}
+
+#[macro_export]
+macro_rules! generate_slot_range_enum {
+	(@
+		$( $parsed:ident )*
+	) => {
+		$crate::generate_slot_range_enum! {
+			/// A compactly represented sub-range from the series.
+			#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, $crate::Encode, $crate::Decode)]
+			#[repr(u8)]
+			pub enum SlotRange { $( $parsed ),* }
+		}
+	};
+	// A trick from StackOverflow to generate the final Enum object.
+	($i:item) => { $i };
```

Comment:
I don't understand why you call recursively
```suggestion
		/// A compactly represented sub-range from the series.
		#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, $crate::Encode, $crate::Decode)]
		#[repr(u8)]
		pub enum SlotRange { $( $parsed ),* }
	};
```